tcptracer-bpf.c

```diff
@@ -0,0 +1,939 @@
+		bpf_probe_read(&dport, sizeof(dport), ((char *)sk) + status->offset_dport);
+		// output
+		struct tcp_ipv4_event_t evt = {
+			.timestamp = bpf_ktime_get_ns(),
+			.cpu = cpu,
+			.type = TCP_EVENT_TYPE_CLOSE,
+			.pid = pid >> 32,
+			.saddr = saddr,
+			.daddr = daddr,
+			.sport = ntohs(sport),
+			.dport = ntohs(dport),
+			.netns = net_ns_inum,
+		};
+		bpf_get_current_comm(&evt.comm, sizeof(evt.comm));
+
+		// do not send event if IP address is 0.0.0.0 or port is 0
+		if (evt.saddr != 0 && evt.daddr != 0 && evt.sport != 0 && evt.dport != 0) {
+			bpf_perf_event_output(ctx, &tcp_event_ipv4, cpu, &evt, sizeof(evt));
+		}
+		struct ipv4_tuple_t t = {
+			.saddr = saddr,
+			.daddr = daddr,
+			.sport = ntohs(sport),
+			.dport = ntohs(dport),
+			.netns = net_ns_inum,
+		};
+		bpf_map_delete_elem(&tuplepid_ipv4, &t);
+	} else if (check_family(sk, AF_INET6)) {
+		u64 saddr_h, saddr_l, daddr_h, daddr_l;
+		saddr_h = 0;
+		saddr_l = 0;
+		daddr_h = 0;
+		daddr_l = 0;
+
+		bpf_probe_read(&daddr_h, sizeof(daddr_h), ((char *)sk) + status->offset_daddr_ipv6);
+		bpf_probe_read(&daddr_l, sizeof(daddr_l), ((char *)sk) + status->offset_daddr_ipv6 + sizeof(u64));
+		bpf_probe_read(&saddr_h, sizeof(saddr_h), ((char *)sk) + status->offset_daddr_ipv6 + 2 * sizeof(u64));
+		bpf_probe_read(&saddr_l, sizeof(saddr_l), ((char *)sk) + status->offset_daddr_ipv6 + 3 * sizeof(u64));
+		bpf_probe_read(&sport, sizeof(sport), ((char *)sk) + status->offset_sport);
+		bpf_probe_read(&dport, sizeof(dport), ((char *)sk) + status->offset_dport);
+
+		struct tcp_ipv6_event_t evt = {
+			.timestamp = bpf_ktime_get_ns(),
+			.cpu = bpf_get_smp_processor_id(),
+			.type = TCP_EVENT_TYPE_CLOSE,
+			.pid = pid >> 32,
+			.saddr_h = saddr_h,
+			.saddr_l = saddr_l,
+			.daddr_h = daddr_h,
+			.daddr_l = daddr_l,
+			.sport = ntohs(sport),
+			.dport = ntohs(dport),
+			.netns = net_ns_inum,
+		};
+		bpf_get_current_comm(&evt.comm, sizeof(evt.comm));
+
+		// do not send event if IP address is :: or port is 0
+		if ((evt.saddr_h || evt.saddr_l) && (evt.daddr_h || evt.daddr_l) && evt.sport != 0 && evt.dport != 0) {
+			bpf_perf_event_output(ctx, &tcp_event_ipv6, cpu, &evt, sizeof(evt));
+		}
+
+		struct ipv6_tuple_t t = {
+			t.saddr_h = saddr_h,
+			t.saddr_l = saddr_l,
+			t.daddr_h = daddr_h,
+			t.daddr_l = daddr_l,
+			t.sport = ntohs(sport),
+			t.dport = ntohs(dport),
+			t.netns = net_ns_inum,
+		};
+		if (is_ipv4_mapped_ipv6(saddr_h, saddr_l, daddr_h, daddr_l)) {
+			struct tcp_ipv4_event_t evt4 = {
+				.timestamp = bpf_ktime_get_ns(),
+				.cpu = bpf_get_smp_processor_id(),
+				.pid = pid >> 32,
+				.type = TCP_EVENT_TYPE_CLOSE,
+				.netns = net_ns_inum,
+				.saddr = (u32)(saddr_l >> 32),
+				.daddr = (u32)(daddr_l >> 32),
+				.sport = ntohs(sport),
+				.dport = ntohs(dport),
+			};
+			bpf_get_current_comm(&evt4.comm, sizeof(evt4.comm));
+			bpf_perf_event_output(ctx, &tcp_event_ipv4, evt4.cpu, &evt4, sizeof(evt4));
```

Comment:
The event is still sent when daddr_l or dport is zero.

The test line 804 is not done here in the ipv4_mapped_ipv6 code path.

@iaguis @alepuccetti